Updated Collected Registration Data

KellenJames opened the pull request:
Updated the frontend and lambda functions to accept a user position and modified the collected expected graduation date.

- The user can pick a position ("Undergraduate", "Graduate", "Faculty/Staff", "Other") when filling out the registration form.
- The expected graduation date input and the major/minor input will only be visible when the "Undergraduate" position is selected.
- The graduation input was modified so that the user can now select a graduating semester and year, instead of a specific date.

I tested these changes in my dev stack cloudfront registration page. I entered the information that is collected when the "Undergraduate" field is chosen, as well as the information when the "Faculty/Staff" field is chosen. The data from both cases was correctly sent to my dev stack dynamoDB table.

Commit:
Updated registration data

Commit:
Updated build artifacts

Commit:
Corrected registration test case

Commit:
Updated to work with the 2 table change

Commit:
Merge branch 'RegistrationData' of https://github.com/KellenJames/unified-makerspace into RegistrationData

KellenJames commented:
This PR should be good to go whenever the data migration is ready, 
I also changed the GradYear back to a string because it would cause errors when the year was left blank because a Faculty/Staff user would leave it blank.

ophilli commented on cdk/visit/console/index.html:
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE html>
```
Are your changes to `cdk/visit/console/` still necessary now that https://github.com/clemsonMakerspace/unified-makerspace/pull/92 has been merged? I think that the pipeline should automatically build the site and we don't need to check in any artifacts now.

ophilli reviewed (changes requested):
Can you rebase this change so that we are prepared to merge?

KellenJames commented:
I rebased it and removed the edits to cdk/visit/console/ , I must have accidently left those in while testing.
I wasn't able to test this change on my dev stack, I'm running into the issue the others are running into with the DynamoDB tables, but it seems like that is being worked on, but I don't believe any of the items changed with this commit should mess with anything I had before.

Commit:
Rebased and tested in Dev

dggorsen commented on cdk/visit/lambda_code/register_user/register_user.py:
```diff
@@ -73,8 +73,10 @@ def add_user_info(self, user_info):
-                'Grad_date': user_info['Grad_Date'],
-                'Major': ', '.join(sorted(user_info['Major'])),
+                'Position': user_info['UserPosition'],
+                'GradSemester': user_info.get('GradSemester', ' '),
+                'GradYear': user_info.get('GradYear', ' '),
```
Are these really optional? Should they be optional based upon "UserPosition"

KellenJames replied:
Yeah they should be optional based upon the UserPosition field, but I didn't think that was something that could be done. I can look into it right now though.

dggorsen commented on cdk/visit/lambda_code/register_user/register_user.py:
```diff
@@ -73,8 +73,10 @@ def add_user_info(self, user_info):
                 'lastName': user_info['lastName'],
```
also can we standardize our attribute casing here?

dggorsen replied:
Oh wait this is the original table- ignore me

dggorsen commented on cdk/visit/lambda_code/register_user/register_user.py:
```diff
@@ -103,8 +105,9 @@ def add_user_info(self, user_info):
                 'last_name': {'S': user_info['lastName']},
                 'gender': {'S': user_info['Gender']},
                 'date_of_birth': {'S': user_info['DOB']},
+                'Position': {'S': user_info['UserPosition']},
```
this should be lowercased

Commit:
Required grad date/major when undergrad is selected

KellenJames commented:
I changed 'Position' to 'position' for the new table.
Grad semester, year, and major are all required now when "Undergraduate Student" is selected, and will not be required from the others.
I tested this on my dev account, and it worked as intended with both undergraduate and graduate selected

dggorsen commented on cdk/visit/lambda_code/register_user/register_user.py:
```diff
@@ -103,8 +105,9 @@ def add_user_info(self, user_info):
+                'position': {'S': user_info['UserPosition']},
                 'grad_semester': {'S': grad_sem},
-                'grad_year': {'N': grad_year},
+                'grad_year': {'S': grad_year},
```
What is the reason for representing this as a string? Bit of a bad practice unless you have a specific reason for it. The reason being, strings require lexicographical sorting and numbers require numerical sorting. Shouldnt really impact us here, but just curious what the rationale was

KellenJames replied:
I made it a string because the registration would fail whenever the year was a number and was left blank by a faculty/staff role, but changing it to a string got it to work.
I'm not sure if my recent change would fix that issue, but I hadn't thought of changing it back.
